pkg/ccl/backupccl/system_schema.go

```diff
@@ -563,16 +563,50 @@ func rekeySystemTable(
 		sort.Sort(toRekey)
 
 		executor := execCtx.InternalExecutor
-		q := strings.Builder{}
-		fmt.Fprintf(&q, "UPDATE %s SET %s = CASE\n", tempTableName, colName)
 
+		// We will update every ID in the table from an old value to a new new value
+		// below, but as we do so there could be yet-to-be-updated rows with old-IDs
+		// at the new IDs which would cause a uniqueness violation before we proceed
+		// to update the row that is in the way. Instead, we will initially offset
+		// the new ID to be +2B, to be up above existing IDs, then when we're done
+		// remapping all of them to their new but offset IDs, we'll slide everything
+		// down to remove the offset at once. We could query the current max, but
+		// just guessing that 2^31 is above it works just as well. In the event this
+		// guess doesn't hold, i.e. a cluster with more than 2^31 descriptors, then
+		// we don't have room in a uint32 OUD for the offset IDs anyway before it
+		// overflows anyway.
+		const offset = 1 << 31
+
+		// Some of the tables use oid columns to store desc IDs rather than ints; in
+		// these tables we will need to cast to int to do the addition/subtraction
+		// of the offset, and then cast back to the desired type determined here.
+		typ := "int"
+		switch tempTableName {
+		case "crdb_temp_system.database_role_settings":
+			typ = "oid"
+		}
+
+		// We'll build one big UPDATE that moves all the IDs to their new values.
```

Comment:
nit: can we also add a line that IDs that are not rekeyed are also moved to a new ID = old ID + offset